use rpms already generated to install hirte components

Yarboa opened the pull request:
Resolve Issue #158

Yarboa committed:
use rpms already generated to install hirte components

Resolve Issue #158
This commit, download artifacts once build workflow complete

It copies the relevant rpm to /tmp/rpms and install it
It also using the man hirte.conf example to create hirte.config files

Signed-off-by: Yariv Rachmani <yrachman@redhat.com>

Yarboa commented on .github/workflows/build.yml:
```diff
@@ -107,3 +107,17 @@ jobs:
+
+  e2e:
+    needs: rpmbuild 
+    runs-on: ubuntu-latest
```
I failed all of my tests to sync jobs from 2 workflows from different yamls.
e2e is running on main if i trigger it manually.

Any idea is welcome
@ygalblum @engelmi

raballew replied:
Something that I have done in the past is inverting the dependencies. So instead of telling the e2e workflow to run once the build workflow is completed, you could directly invoke the e2e workflow from the build workflow.

This can be done by something like this:

**.github/workflows/e2e.yml**
```yml
name: Integration Testing
on: [workflow_call] 
...
```

**.github/workflows/deploy.yml**
```yml
name: Build
jobs:
  rpmbuild:
    ...
  e2e:
    uses: ./.github/workflows/e2e.yml  
    needs: rpmbuild
```

Yarboa replied:
Thanks, 
I did not try this one,

raballew commented on .github/workflows/e2e.yaml:
```diff
@@ -1,17 +1,24 @@
+  workflow_run:
+    workflows: ["Build"]
+    types: 
+      - completed
```
You might want to run the integration tests only if the build workflow completed successfully: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-a-workflow-based-on-the-conclusion-of-another-workflow

Yarboa replied:
Ack, will give the whole a try

Yarboa committed:
Sync simply buildrpm job with e2e

build.yml triggers e2e directly
rename e2e.yaml -> e2e.yml for consistency

Signed-off-by: Yariv Rachmani <yrachman@redhat.com>

mwperina commented on .github/workflows/build.yml:
```diff
@@ -109,3 +109,8 @@ jobs:
+
+  e2e:
+    needs: rpmbuild
+    uses: ./.github/workflows/e2e.yml
```
Can't we use relative path?

```
    uses: e2e.yml
```

Yarboa replied:
No, Please refer this doc
https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/reusing-workflows#example-caller-workflow

mwperina replied:
OK

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+          podman exec -it hirte-mgr bash -c "man hirte.conf | \
+               grep -A 5 \"\[hirte\]\" | sed -e 's/^ *//g' \
+               -e 's/journald/stderr/' \
+               -e 's/agent.*/foo,bar/' > /etc/hirte/hirte.conf"
```
I'd really prefer to clearly define what's changed here. So instead of relying on man page content, I'd update specific options, for example

```
# Allow nodes foo and bar to connect
          podman exec -it hirte-mgr \
              sed -i "s/\(AllowedNodeNames=\)\(.*\)/\1foo,bar/" /etc/hirte/hirte.conf
          # Change LogLevel to DEBUG
          podman exec -it hirte-mgr \
              sh -c 'echo "LogLevel=DEBUG" >> /etc/hirte/hirte.conf'
```

Yarboa replied:
Ack, but currently /etc/hirte/hirte.conf is empty

mwperina replied:
It's not empty, it should contain https://github.com/containers/hirte/blob/main/config/hirte/hirte.conf after hirte RPM is installed

Yarboa replied:
Ack

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+               -e 's/journald/stderr/' \
+               -e 's/agent.*/foo,bar/' > /etc/hirte/hirte.conf"
+          podman exec -d  hirte-mgr \
+               bash -c "hirte > mgr.log 2>&1 &"
```
Can't we use systemctl to start services?

```
podman exec -d systemctl start hirte
```

Also wouldn't it be better to use `journactl` instead of redirecting stderr to manual log file?

Yarboa replied:
Ack, for journal, related to systemctl, does the unit files ready by the rpm install?

Prefer to have it, once the bits are installable

mwperina replied:
It's already included in RPM installation: https://github.com/containers/hirte/pull/181

Yarboa replied:
OK it is new, thanks

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+          podman exec -it hirte-node-bar bash -c "man hirte-agent.conf | \
+               grep -A 6 \"\[hirte-agent\]\" | sed -e 's/^ *//g' \
+               -e 's/journald/stderr/' -e \"s/127.0.0.1/${{ env.PODMAN_NET }}/\" \
+               -e 's/agent-.*/bar/' > /etc/hirte/agent.conf"
```
Same as above, I'd prefer well described changes:

```
# Change node name in config file
podman exec -it hirte-agent-bar \
    sed -i "s/\(NodeName=\)\(.*\)/\1bar/" /etc/hirte/agent.conf
# Change manager address in config file
podman exec -it hirte-agent-bar \
     sed -i "s/\(ManagerHost=\)\(.*\)/\1${{ env.PODMAN_NET }}/" /etc/hirte/agent.conf
# Change LogLevel to DEBUG
podman exec -it hirte-agent-bar \
    sh -c 'echo "LogLevel=DEBUG" >> /usr/local/etc/hirte/hirte.conf'
```

Yarboa replied:
Ack

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+          podman exec -d hirte-node-bar \
+               bash -c "pkill --signal TERM hirte-agent"
+          podman exec -it hirte-node-bar bash -c \
+               "grep \"Agent 'bar' connection attempt 0: success\" bar.log"
```
Can't we use journalctl instead of manual log file?

```
podman exec -it hirte-agent-bar \
              sh -c 'journalctl | grep "connection attempt 0: success" | grep bar'
```

Yarboa replied:
Ack

mwperina commented on tests/Containerfile:
```diff
@@ -1,11 +1,9 @@
+COPY ./exported-artifacts/*.rpm /tmp/rpms/
+RUN /usr/bin/find /tmp/rpms -name hirte-[0-9].*.x86_64.rpm -o \
+                            -name hirte-agent-[0-9].*.x86_64.rpm | \
+                            xargs  dnf install -y
```
Wouldn't it be easier to reuse repository included in `rpm-artifacts.zip`?

```
dnf install -y \
    --repo local-hirte-snapshot \
    --repofrompath local-hirte-snapshot,file:///tmp/rpms \
    hirte hirte-agent
```

Yarboa replied:
Ack, better

raballew commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+
+      - name: run hirte agent-foo
+        run: |
+          podman exec -it hirte-node-foo bash -c "man hirte-agent.conf | \
```
Inconsistent naming: shouldn't it be either hirte-agent or hirte-node. Even the log uses the Agent terminology.

mwperina replied:
hirte-agent is running on a node, I don't see an inconsistency here

raballew commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+
+      - name: run hirte agent-bar
+        run: |
+          podman exec -it hirte-node-bar bash -c "man hirte-agent.conf | \
```
Inconsistent naming: shouldn't it be either hirte-agent or hirte-node. Even the log uses the Agent terminology.

mwperina replied:
Same as above

Yarboa replied:
Ack

raballew commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,80 @@
+
+      - name: run hirte orchestrator
+        run: |
+          podman exec -it hirte-mgr bash -c "man hirte.conf | \
```
Inconsistent naming: Is it the hirte orchestrator or hirte-mgr?

Yarboa replied:
Ack

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,98 @@
+        run: |
+          podman exec -it hirte-mgr \
+               bash -c "cat <<EOF>> /etc/hirte/hirte.conf
+               ManagerPort=2020
```
Not needed, default value

Yarboa replied:
@mwperina 
Thanks martin, 
Based on the following:
* hirte.conf, there are no parameter names at all.
* man hirte.conf, no defaults mentioned in it

This is way I used the man values in the beginning

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,98 @@
+               ManagerPort=2020
+               AllowedNodeNames=foo,bar
+               LogLevel=DEBUG
+               LogTarget=journald
```
Not needed, default value

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,98 @@
+               AllowedNodeNames=foo,bar
+               LogLevel=DEBUG
+               LogTarget=journald
+               LogIsQuiet=false
```
Not needed, default value

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,98 @@
+               ManagerPort=2020
+               LogLevel=DEBUG
+               LogTarget=journald
+               LogIsQuiet=false
```
Same as above, no need to change default values

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,98 @@
+               ManagerPort=2020
+               LogLevel=DEBUG
+               LogTarget=journald
+               LogIsQuiet=false
```
Same as above, no need to change default values

Yarboa committed:
Update e2e to work hirte unit files

build.yml triggers e2e directly
rename e2e.yaml -> e2e.yml for consistency
Work with systemd unitfiles and journalctl

Signed-off-by: Yariv Rachmani <yrachman@redhat.com>

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,89 @@
+          LogLevel=DEBUG
+          _EOF"
+          podman exec -d hirte-node-bar \
+               bash -c "systemctl enable hirte-agent.service; systemctl start hirte-agent.service"
```
No need to enable the service, we will not test service status after a reboot

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,89 @@
+          LogLevel=DEBUG
+          _EOF"
+          podman exec -it hirte-mgr \
+               bash -c "systemctl enable hirte.service; systemctl start hirte.service"
```
No need to enable the service, we will not test service status after a reboot

mwperina commented on .github/workflows/e2e.yml:
```diff
@@ -0,0 +1,89 @@
+          LogLevel=DEBUG
+          _EOF"
+          podman exec -d hirte-node-foo \
+               bash -c "systemctl enable hirte-agent.service; systemctl start hirte-agent.service"
```
No need to enable the service, we will not test service status after a reboot

mwperina reviewed (changes requested):
Some small comments, but other than that it looks good

mwperina reviewed (approved):
+1

engelmi reviewed (approved):
LGTM
